framework/src/media/audio/audio_manager.c

```diff
@@ -725,21 +786,40 @@ audio_manager_result_t set_audio_stream_in(unsigned int channels, unsigned int s
 		goto error_with_pcm;
 	}
 
+	card->resample.format = pcm_format_to_bits(pcm_get_format(card->pcm)) >> 3;	// ToDo: Change into the argument "format".
+
+	card->resample.user_channel = channels;
+	rechanneling_ratio = (float)card->resample.user_channel / (float)config.channels;
+	if (rechanneling_ratio < 1) {
+		card->resample.rechanneling_buffer_size = get_user_input_frames_to_byte(get_input_frame_count()) / rechanneling_ratio;
+		card->resample.rechanneling_buffer = malloc(card->resample.rechanneling_buffer_size);
+		medvdbg("Rechanneling ratio = %f, rechanneling buffer size = %d\n", rechanneling_ratio, card->resample.rechanneling_buffer_size);
+	}
+
 	if (sample_rate == config.rate) {
 		card->resample.necessary = false;
 	} else {
+		float resample_buffer_size = (float)get_user_input_frames_to_byte(get_input_frame_count());
+
 		card->resample.necessary = true;
 		card->resample.from = config.rate;
 		card->resample.to = sample_rate;
 		card->resample.ratio = (float)sample_rate / (float)config.rate;
 		card->resample.handle = src_init(CONFIG_AUDIO_RESAMPLER_BUFSIZE);
-		card->resample.buffer = malloc((int)((float)get_input_frames_to_byte(get_input_frame_count()) / card->resample.ratio) + 1);	// +1 for floating point margin
+
+		resample_buffer_size /=  card->resample.ratio;
+		if (resample_buffer_size - (int)resample_buffer_size > 0) {
```

Comment:
Suppose that we are resample 20K to 48K, 16bit and stereo and buffer size is 4096.
In this case ratio is 2.4
so buffer size will be calculated as 9830.4 -> 9831.
Let's think about how many frames can be stored in 9831 bytes.
4096 bytes, we can store 1024 frames.
This 1024 frames should be changed to 2457.6 frames (20k->48k), 2458 frames.
2458 frames * 4 = 9832. This means we lost last 1 bytes 
So instead of +1, we have to calculate based on bit rate & channel here.